File: php-ffi/cli_size/cli_size.c
/**
 * 通过函数 ioctl(); 获得终端界面的参数
 * @see https://blog.csdn.net/weixin_42205987/article/details/82080615
 */

//具体实现方法
#include <stdio.h>
#include <sys/ioctl.h>
#include <unistd.h>

unsigned short get_cli_rows()
{
    //定义winsize 结构体变量
    struct winsize size;
    //TIOCSWINSZ命令可以将此结构的新值存放到内核中
    ioctl(STDIN_FILENO, TIOCGWINSZ, &size);
    // printf("%d\n", size.ws_col);
    // printf("%d\n", size.ws_row);
    // return 0;
    return size.ws_row;
}

unsigned short get_cli_cols()
{
    //定义winsize 结构体变量
    struct winsize size;
    //TIOCSWINSZ命令可以将此结构的新值存放到内核中
    ioctl(STDIN_FILENO, TIOCGWINSZ, &size);
    // printf("%d\n", size.ws_col);
    // printf("%d\n", size.ws_row);
    // return 0;
    return size.ws_col;
}

struct winsize get_size()
{
    //定义winsize 结构体变量
    struct winsize size;
    //TIOCSWINSZ命令可以将此结构的新值存放到内核中
    ioctl(STDIN_FILENO, TIOCGWINSZ, &size);
    return size;
}

File: php-ffi/cli_size/cli_size.php
<?php

// $res = exec("./a.out", $output, $return_var);
// var_dump($res, $output, $return_var);

// gcc -c -fPIC -o cli_size.o cli_size.c
// gcc -shared -o libcli_size.so cli_size.o
// or gcc -O2 -fPIC -shared -g cli_size.c -o libcli_size.so

$ffi = FFI::cdef(<<<CTYPE
struct winsize {
	unsigned short  ws_row;         /* rows, in characters */
	unsigned short  ws_col;         /* columns, in characters */
	unsigned short  ws_xpixel;      /* horizontal size, pixels */
	unsigned short  ws_ypixel;      /* vertical size, pixels */
};
unsigned short get_cli_rows(); 
unsigned short get_cli_cols(); 
struct winsize get_size();
CTYPE, 'libcli_size.so');
//var_dump($ffi);
var_dump($ffi->get_cli_rows());
var_dump($ffi->get_cli_cols());
var_dump($ffi->get_size());
var_dump($ffi->get_size()->ws_row);
